### PR TITLE
Rewrote BoxUniform such that it can be used as a base for a flow

### PR DESCRIPTION
**Pull request by rbvh**
Hello,

Back again trying to contribute something useful. I found that if I use BoxUniform as the base distribution for a flow model I get 

`log_prob() got an unexpected keyword argument 'context'`

I rewrote BoxUniform to look more like StandardNormal to prevent that issue. I could also rename it if the original BoxUniform needs to stay around. 

Cheers,
Rob

### CODE REVIEW

**Comment by sambklein** [comment]
Could this or something like it please be accepted

File: nflows/distributions/uniform.py [hunk 1]
```diff
@@ -3,32 +3,61 @@
...
+                "low has elements that are higher than high"
+            )
+
+        self._shape = low.shape
+        self._low = low
+        self._high = high
+        self._log_prob_value = -torch.sum(torch.log(high - low))
+
+    def _log_prob(self, inputs, context):
+        # Note: the context is ignored.
```
Comment (sitmo) [inline nflows/distributions/uniform.py:41]: Can we make the context set new lower/upper bounds INSIDE the current bounds? I.e. let the context select a subregion of the domain? context_low = low + (high-low)*sigmoid(context)?

**Comment by MilesCranmer** [comment]
+1

See #33 which is a more general fix for this as well as for other `torch.distributions.Independent` objects

File: nflows/distributions/uniform.py (hunk 1 above)
Comment (MilesCranmer) [reply nflows/distributions/uniform.py:41]: also really like this idea^. Maybe a ConditionalUniform class?

Comment (MilesCranmer) [reply nflows/distributions/uniform.py:41]: Although if a point is outside the distribution, things will blow up. Maybe useful to have another distribution with a softer drop outside the uniform bounds?

Comment (sitmo) [reply nflows/distributions/uniform.py:41]: Yes, I believe too that the conditional distribution should be inside the unconditional one, and that's best done by picking a transform that guarantees it. I think one way to argue for having a conditional distribution that is a box inside the original box is via Bayes theorem for conditional distributions P(A|B) = P(A & B) / P(B). The P(A & B) is the intersection of the original box with a new box, which is a sub-box? But maybe it's not a very good concept to use in this case.

I also like your other idea of allowing flows to use base distributions that don't have a context. Another way to do that is by making the context argument optional in the flow's base class,  "_log_prob(self, inputs, context**=None**)". And then forward it like self._distribution.log_prob(noise) only if context is None. That way you can use use the distributions that don't have a context in flows, but only if the flow doesn't have a context either. This might prevent people from accidentally using a non-context distribution with a flow that does have a context?

Btw, do you know if we can use Splines instead of Afine in the conditional moons example? I tried it, but I get errors. It perhaps look like splines don't work with conditional Gaussian distributions as a base distribution? Do splines require distributions with a finite domain? If so, then having a conditional uniform distribution might make it possible?

Comment (MilesCranmer) [reply nflows/distributions/uniform.py:41]: I did get the spline code working at some point, I don't remember if I made changes but I don't think you need to.

Re: the context, actually it should work with a context-less base distribution and context-full transformations! This is what I need this modification for actually.

**Comment by sitmo** [comment]
Ah cool, thanks! I was wondering if it were possible to have have context
dependent transforms. I thought the only way to providing a context was via
the base distribution!
I'm going to try that now!

**Comment by MilesCranmer** [comment]
No problem!

The conditional moons example actually does both context for the base distribution AND the transform:
```python
base_dist = ConditionalDiagonalNormal(shape=[2], 
                                      context_encoder=nn.Linear(1, 4))

transforms = []
for _ in range(num_layers):
    transforms.append(ReversePermutation(features=2))
    transforms.append(MaskedAffineAutoregressiveTransform(features=2, 
                                                          hidden_features=4, 
                                                          context_features=1))
transform = CompositeTransform(transforms)

flow = Flow(transform, base_dist)
```
^ see the `context_features` kwarg to MaskedAffineAutoregressiveTransform? That is actually the conditional feature for the actual distribution transform.

Cheers,
Miles

**Comment by sitmo** [comment]
Thanks Miles, that’s very helpful, I'm starting to better understand the nuts and bolts now!

**Comment by MilesCranmer** [comment]
Update: I'm not sure when #33 will get merged but you can install the fixed version with:
```python
pip install -U git+https://github.com/MilesCranmer/nflows.git@dev
```
This should fix these errors. I'll try to keep it up-to-date with the master branch until things get merged.
